lib/Conversion/ExportVerilog/HWLowerInstanceChoices.cpp

```diff
@@ -0,0 +1,126 @@
+//===- HWLowerInstanceChoices.cpp - IR Prepass for Emitter ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file implements the "prepare" pass that walks the IR before the emitter
+// gets involved.  This allows us to do some transformations that would be
+// awkward to implement inline in the emitter.
+//
+// NOTE: This file covers the preparation phase of `ExportVerilog` which mainly
+// legalizes the IR and makes adjustments necessary for emission. This is the
+// place to mutate the IR if emission needs it. The IR cannot be modified during
+// emission itself, which happens in parallel.
+//
+//===----------------------------------------------------------------------===//
+
+#include "../PassDetail.h"
+#include "ExportVerilogInternals.h"
+#include "circt/Conversion/ExportVerilog.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Support/LLVM.h"
+#include "circt/Support/Namespace.h"
+#include "circt/Support/SymCache.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/Support/Debug.h"
+
+#include <mutex>
+
+#define DEBUG_TYPE "lower-hw-instance-choices"
+
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+using namespace sv;
+using namespace ExportVerilog;
+
+namespace {
+
+struct HWLowerInstanceChoicesPass
+    : public HWLowerInstanceChoicesBase<HWLowerInstanceChoicesPass> {
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+LogicalResult
+ExportVerilog::lowerHWInstanceChoices(mlir::ModuleOp module,
+                                      const LoweringOptions &options) {
+  // Collect all instance choices & symbols.
+  SmallVector<InstanceChoiceOp> instances;
+  SymbolCache symCache;
+  {
```

Comment:
Interesting use of a scope here. Normally I expect this if a PR is trying to hide variables from the outer scope (as the PR does with `os` later which looks good!). What is this accomplishing here? Is this trying to say that everything in the scope is related to the initialization of the previous two variables?

Reply:
Useless now, but if there are multiple constructs, I do try to isolate them under the shared data they build.